### ISSUE TITLE
CGAL::do_intersect(Segment,Segment) breaks on Visual Studio 2017

### ISSUE THREAD

**Issue by justinjk007** [CGAL/cgal#2413]
_Please use the following template to help us solving your issue._

## Issue Details

I recently switched from Visual Studio 2015 to 17. My code which was running without any problem until then on both linux and Windows breaks when ran after compiling in Visual Studio 2017. Very strange behavior.  Only `CGAL::do_intersect(Segment,Segment)` have problems. I am using many other functions to tackle polygons and do other stuff with Segments and so on. 

## Source Code

``` c++
bool doIntersect(Segment line, std::list<Segment> lines, Segment except_line) {
  /**
   * Check if the passed line segments intersects with any of the
   * lines inside the given list of lines without checking for the
   * line that is an exception.
   */

  for (std::list<Segment>::iterator it = lines.begin(); it != lines.end(); it++)
    if(*it == except_line) {
      continue;
    }
    else {
      if(CGAL::do_intersect(line , *it)) // Breaks here when run, compiles fine.
    	return true;
    }
  return false;
}
```
## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): Windows 10
* Compiler: MS VSC-2017
* Release or debug mode: debug
* CGAL version: 4.10
* Boost version: 1.69

This is very weird. I tried digging into CGAL source code but I am only a beginner and got lost very easily .  For the time being I implemented a function for myself to check if two liners are intersecting . I am just letting you guys know.

**Comment by sloriot** [comment]
What kernel did you use?

**Comment by justinjk007** [comment]
Oh my bad, these are the declarations for cgal.
```c++
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Polygon_2.h>
#include <CGAL/Simple_cartesian.h>
#include <CGAL/algorithm.h>
#include <CGAL/bounding_box.h>
#include <fstream>
#include <iostream>
#include <list>
#include "myGeometry.hpp"

typedef CGAL::Exact_predicates_inexact_constructions_kernel K;

typedef K::FT FT;
typedef K::Point_2 Point_2;
typedef K::Segment_2 Segment;
typedef K::Iso_rectangle_2 Rect;
typedef CGAL::Polygon_2<K> Polygon_2;

```

**Comment by sloriot** [comment]
Could you try to spot a pair of segments causing the issue? Our testsuite did not have any pb with MSVC 17 on the two platforms we have. Thanks.

**Comment by justinjk007** [comment]
Sorry for the delay, totally forgot about this.

Ok I tested the following code on both the VS2015 and VS2017.

This time both compilers ran perfectly fine. But In my big project it still breaks when i'm on VS 2017, but runs fine with VS2015. I don't have the skill or time to go looking forward to find that `Segment` which breaks the method. But my implementation do the job for me now. I will however trace the bug in few weeks time when I find free time to trace that `Segment` down. 

In the **main.cpp** I had this

```c++
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Polygon_2.h>
#include <CGAL/Simple_cartesian.h>
#include <CGAL/algorithm.h>
#include <CGAL/bounding_box.h>
#include <iostream>

typedef CGAL::Exact_predicates_inexact_constructions_kernel K;

typedef K::FT FT;
typedef K::Point_2 Point;
typedef K::Segment_2 Segment;
typedef K::Iso_rectangle_2 Rect;
typedef CGAL::Polygon_2<K> Polygon_2;

int main()
{
    Point a1(9.3, 5.5);
    Point b1(9.3, 5.5);

    Point a2(9.3, 5.5);
    Point b2(9.3, 5.5);

    Segment one(a1, b1);
    Segment two(a2, b2);

    if (CGAL::do_intersect(one, two))
        std::cout << "These lines intersect" << std::endl;  // This is correct
    else
        std::cout << "These lines doesn't intersect" << std::endl;

    return 0;
}

```

And this was my **CMakeLists.txt**

```cmake
CMAKE_MINIMUM_REQUIRED (VERSION 3.5)
PROJECT (CGALBugRecreation)

SET (CMAKE_CXX_STANDARD 11)

# CGAL config
FIND_PACKAGE(CGAL REQUIRED)
INCLUDE( ${CGAL_USE_FILE} )
INCLUDE( CGAL_CreateSingleSourceCGALProgram )
INCLUDE_DIRECTORIES ("${CGAL_ROOT}/include/CGAL")
INCLUDE_DIRECTORIES (BEFORE "../../include")

ADD_EXECUTABLE(main main.cpp)
```

**Comment by janetournois** [comment]
I have tried the above example with CGAL-4.10 and CGAL-4.10.1, and it worked fine in both Debug and Release modes, on vs2017.
The only difference I see is that our boost version is 1.63 (not sure it has any impact though).

Is there a chance that you are using a different version of CGAL with your vs2015 and vs2017?

**Comment by justinjk007** [comment]
@janetournois The above code runs fine on both VS2015 and VS2017, for me too. But he code that was part of my big project breaks when it is searching through a list of segments for intersection. But the same code runs fine on VS2015 and also when using G++ 6.x. 

I will investigate which segment breaks the code and get back to you.

**Comment by janetournois** [comment]
ok, thanks

**Comment by justinjk007** [comment]
OK I have instigated the issue. I think its not an issue with CGAL itself its some sort of memory leak happening in my program . I isolated the method that was causing the problem and it runs fine after processing 400 segments, which is when it breaks on my main program. So definitely  my fault.

I have not really taken care of memory management in my program yet.  I will do that and may be tell you if that fixed it. Thanks for sticking around.
